File: TopInterview150/Stack/EvaluateReversePolishNotation/solution.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        int operand1;
        int operand2;
        for (string token : tokens)
        {
            if (token == "*" || token == "/" || token == "+" || token == "-")
            {
                operand2 = st.top();
                st.pop();
                operand1 = st.top(); st.pop();
                if (token == "/")
                    st.push(operand1/operand2);
                else if (token == "*")
                    st.push(operand1*operand2);
                else if (token == "-")
                    st.push(operand1-operand2);
                else
                    st.push(operand1+operand2);
            }
            else st.push(stoi(token));
        }
        return st.top();
    }
};

File: TopInterview150/Stack/EvaluateReversePolishNotation/solution.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []
        operand1 = 0
        operand2 = 0
        for token in tokens:
            if token in ["*", "-", "/", "+"]:
                operand2 = stack.pop()
                operand1 = stack.pop()
                if token == "+":
                    stack.append(operand1+operand2)
                elif token == "-":
                    stack.append(operand1-operand2)
                elif token == "/":
                    result = int(operand1 / operand2) if (operand1/operand2 >= 0) else -(abs(operand1)//abs(operand2))
                    stack.append(result)
                else:
                    stack.append(operand1*operand2)
            else:
                stack.append(int(token))
        return stack.pop()
